File: mobile/ios/HADEOneScreen/Models/SpontaneousObject.swift
import Foundation

enum SpontaneousObjectType: String, Codable {
    case ugcEvent = "ugc_event"
    case placeOpportunity = "place_opportunity"
}

enum SpontaneousUserState: String, Codable {
    case going = "going"
    case maybe = "maybe"
}

struct SpontaneousTimeWindow: Codable, Equatable {
    let start: Double
    let end: Double
}

struct SpontaneousLocation: Codable, Equatable {
    let lat: Double
    let lng: Double
    let placeId: String?

    enum CodingKeys: String, CodingKey {
        case lat, lng
        case placeId = "place_id"
    }
}

struct SpontaneousObject: Codable, Identifiable, Equatable {
    let id: String
    let type: SpontaneousObjectType
    let title: String
    let timeWindow: SpontaneousTimeWindow
    let location: SpontaneousLocation
    let radius: Double
    let goingCount: Int
    let maybeCount: Int
    let userState: SpontaneousUserState?
    let createdAt: Double
    let expiresAt: Double
    let trustScore: Double
    let vibeTag: String?
    let source: String?

    enum CodingKeys: String, CodingKey {
        case id, type, title, radius, source
        case timeWindow = "time_window"
        case location
        case goingCount = "going_count"
        case maybeCount = "maybe_count"
        case userState = "user_state"
        case createdAt = "created_at"
        case expiresAt = "expires_at"
        case trustScore = "trust_score"
        case vibeTag = "vibe_tag"
    }

    static func fromGooglePlace(
        id: String,
        name: String,
        lat: Double,
        lng: Double,
        placeId: String? = nil,
        vibeTag: String? = nil
    ) -> SpontaneousObject {
        let now = Date().timeIntervalSince1970
        return SpontaneousObject(
            id: id,
            type: .placeOpportunity,
            title: name,
            timeWindow: SpontaneousTimeWindow(start: now, end: now + 7200),
            location: SpontaneousLocation(lat: lat, lng: lng, placeId: placeId),
            radius: 500,
            goingCount: 0,
            maybeCount: 0,
            userState: nil,
            createdAt: now,
            expiresAt: now + 7200,
            trustScore: 0.5,
            vibeTag: vibeTag,
            source: "google_places"
        )
    }

    static func fromUGC(
        id: String,
        title: String,
        lat: Double,
        lng: Double,
        placeId: String? = nil,
        timeWindowStart: Double? = nil,
        timeWindowEnd: Double? = nil,
        radius: Double = 300,
        goingCount: Int = 0,
        maybeCount: Int = 0,
        userState: SpontaneousUserState? = nil,
        createdAt: Double? = nil,
        expiresAt: Double? = nil,
        trustScore: Double = 0.5,
        vibeTag: String? = nil,
        source: String? = nil
    ) -> SpontaneousObject {
        let now = Date().timeIntervalSince1970
        return SpontaneousObject(
            id: id,
            type: .ugcEvent,
            title: title,
            timeWindow: SpontaneousTimeWindow(
                start: timeWindowStart ?? now,
                end: timeWindowEnd ?? now + 7200
            ),
            location: SpontaneousLocation(lat: lat, lng: lng, placeId: placeId),
            radius: radius,
            goingCount: goingCount,
            maybeCount: maybeCount,
            userState: userState,
            createdAt: createdAt ?? now,
            expiresAt: expiresAt ?? now + 7200,
            trustScore: trustScore,
            vibeTag: vibeTag,
            source: source
        )
    }

    func updateParticipation(_ newState: SpontaneousUserState?) -> SpontaneousObject {
        if userState == newState { return self }

        var nextGoingCount = goingCount
        var nextMaybeCount = maybeCount

        if userState == .going {
            nextGoingCount = max(0, nextGoingCount - 1)
        }
        if userState == .maybe {
            nextMaybeCount = max(0, nextMaybeCount - 1)
        }

        if newState == .going {
            nextGoingCount += 1
        }
        if newState == .maybe {
            nextMaybeCount += 1
        }

        return SpontaneousObject(
            id: id,
            type: type,
            title: title,
            timeWindow: timeWindow,
            location: location,
            radius: radius,
            goingCount: nextGoingCount,
            maybeCount: nextMaybeCount,
            userState: newState,
            createdAt: createdAt,
            expiresAt: expiresAt,
            trustScore: trustScore,
            vibeTag: vibeTag,
            source: source
        )
    }

    func generateExplanation() -> String {
        let now = Date().timeIntervalSince1970
        let start = normalizedSeconds(timeWindow.start)
        let minutesUntilStart = max(0, Int(ceil((start - now) / 60)))
        let timeCopy = start <= now ? "happening now" : "starting in \(minutesUntilStart) min"
        let participationCopy = goingCount == 1 ? "1 person is going" : "\(goingCount) people are going"

        return "This is \(timeCopy), and \(participationCopy)."
    }

    private func normalizedSeconds(_ timestamp: Double) -> Double {
        timestamp > 10_000_000_000 ? timestamp / 1000 : timestamp
    }
}

File: domain/spontaneous-object/spontaneousObject.ts
export type SpontaneousObjectType = "ugc_event" | "place_opportunity";
export type UserState = "going" | "maybe" | null;

export interface SpontaneousObject {
  id: string;
  type: SpontaneousObjectType;
  title: string;
  time_window: { start: number; end: number };
  location: { lat: number; lng: number; place_id?: string };
  radius: number;
  going_count: number;
  maybe_count: number;
  user_state: "going" | "maybe" | null;
  created_at: number;
  expires_at: number;
  trust_score: number;
  vibe_tag?: string;
  source?: string;
}

/** Raw Google Places API response shape accepted by fromGooglePlace. */
export interface GooglePlaceInput {
  place_id: string;
  name: string;
  geometry: { location: { lat: number; lng: number } };
  vibe?: string;
}

export interface UGCInput {
  id: string;
  title: string;
  location: { lat: number; lng: number; place_id?: string };
  type?: SpontaneousObjectType;
  time_window?: { start?: number; end?: number };
  radius?: number;
  going_count?: number;
  maybe_count?: number;
  user_state?: UserState;
  trust_score?: number;
  vibe_tag?: string;
  source?: string;
  created_at?: number;
  expires_at?: number;
}

export function fromGooglePlace(place: GooglePlaceInput): SpontaneousObject {
  const now = Date.now();
  return {
    id: place.place_id,
    type: "place_opportunity",
    title: place.name,
    time_window: { start: now, end: now + 60 * 60 * 1000 },
    location: { lat: place.geometry.location.lat, lng: place.geometry.location.lng, place_id: place.place_id },
    radius: 500,
    going_count: 0,
    maybe_count: 0,
    user_state: null,
    created_at: now,
    expires_at: now + 2 * 60 * 60 * 1000,
    trust_score: 0.5,
    vibe_tag: place.vibe,
    source: "google_places",
  };
}

export function fromUGC(input: UGCInput): SpontaneousObject {
  const now = Date.now();
  return {
    id: input.id,
    type: "ugc_event",
    title: input.title,
    time_window: {
      start: input.time_window?.start ?? now,
      end: input.time_window?.end ?? now + 2 * 60 * 60 * 1000,
    },
    location: input.location,
    radius: input.radius ?? 300,
    going_count: input.going_count ?? 0,
    maybe_count: input.maybe_count ?? 0,
    user_state: input.user_state ?? null,
    created_at: input.created_at ?? now,
    expires_at: input.expires_at ?? now + 2 * 60 * 60 * 1000,
    trust_score: input.trust_score ?? 0.5,
    vibe_tag: input.vibe_tag,
    source: input.source,
  };
}

export function updateParticipation(
  object: SpontaneousObject,
  newState: UserState,
): SpontaneousObject {
  if (object.user_state === newState) return object;

  let going = object.going_count;
  let maybe = object.maybe_count;

  if (object.user_state === "going") going = Math.max(0, going - 1);
  if (object.user_state === "maybe") maybe = Math.max(0, maybe - 1);

  if (newState === "going") going += 1;
  if (newState === "maybe") maybe += 1;

  return { ...object, going_count: going, maybe_count: maybe, user_state: newState };
}

export function generateExplanation(object: SpontaneousObject): string {
  const now = Date.now();
  const minutesUntilStart = Math.max(0, Math.ceil((object.time_window.start - now) / 60_000));
  const timeCopy = object.time_window.start <= now ? "happening now" : `starting in ${minutesUntilStart} min`;
  const participationCopy = object.going_count === 1 ? "1 person is going" : `${object.going_count} people are going`;
  return `This is ${timeCopy}, and ${participationCopy}.`;
}
